### ISSUE TITLE
CS231n_Loss_Function_and_Optimizatiion

### ISSUE THREAD

**Issue by wangchau** [wangchau/blog#12]
# Loss Function and Optimizatiion

- 写一个算法，自动决定哪些W是最优的，就需要一个度量任意某个W的好坏的方法，定义一个函数把W当作输入，然后看一下得分，定量的估计W的好坏，这个函数称为损失函数。
- 定量的衡量任意一个W的好坏，找到一个有效的方法来从W的可行域里找到W的取值，在不最坏的情况。这个过程称为优化过程。

## 损失函数

![screen shot 2018-01-22 at 11 16 47](https://user-images.githubusercontent.com/35488201/35309891-90f159d2-00e9-11e8-837b-6ebbeca2603d.png)

- 有N个样本，x是算法的输入，在图像分类问题里，x是每个像素点所构成的数据集，y是希望算法预测出来的东西，我们称之为标签或者目标。在这个问题里，y是1-10（or 0-9）之间的整数，以为我们我们总共有10种分类。这个整数表明，对每个图片的哪个类是正确的。能够得出参数W是否令人满意。我们可以找出在训练集上损失函数最小化的W。

### 多分类SVM损失函数

![screen shot 2018-01-22 at 12 26 42](https://user-images.githubusercontent.com/35488201/35309899-9adc3412-00e9-11e8-913f-e743dccc55d0.png)

- 多分类SVM是在处理多分类时候的一种推广。
- 损失函数的目的：这个损失是在说，如果真实分类的分数比其他的分数高很多，就会很满意，但是要高出多少，就定义了一个安全的边距，如果小于这个边距，我们会得到一些损失。
- 对于这里所说的这个问题，$x_i$的像素和代表正确分类的标签$y_i$。评分函数输入像素数据，然后通过公式 $f(x_i,W)$来计算不同分类类别的分值。这里我们记分值为s。比如针对地j个类别的得分就是第j个元素：$s_j = f(x_i,W)_j$。针对第i 个数据的分类SVM的损失函数定义如下图所示。

#### 折叶函数

- 这里，形如$max(0,-)$的函数，我们称为折叶函数（hinge loss）。有时候会听到人们使用平方折叶损失SVM（即L2-SVM），它使用的是$max(0,-)^2$将更强烈（平方地而不是线性地）地惩罚过界的边界值。

![screen shot 2018-01-22 at 15 05 37](https://user-images.githubusercontent.com/35488201/35309903-a4eff312-00e9-11e8-8f3c-d25f6c454bef.png)

- 这里我们将∆定为1，则折叶函数的形式如图所示：

<img width="881" alt="screen shot 2018-01-22 at 15 12 45" src="https://user-images.githubusercontent.com/35488201/35309905-ab70845e-00e9-11e8-90df-bbab545105a7.png">

- 多类SVM“想要”正确类别的分类分数比其他不正确分类类别的分数要高，而且至少高出delta的边界值。如果其他分类分数进入了红色的区域，甚至更高，那么就开始计算损失。如果没有这些情况，损失值为0。我们的目标是找到一些权重，它们既能够让训练集中的数据样例满足这些限制，也能让总的损失值尽可能地低。

#### Example

![screen shot 2018-01-22 at 15 14 40](https://user-images.githubusercontent.com/35488201/35309911-b28c0812-00e9-11e8-8ff9-0fea84d3ba21.png)

- 如上图所示，我们这里设∆ = 1。这幅图片的分类是cat，$s_(yi)$ = 3.2,这里的$s_j$分别为5。1和-1.7，所以按照公式计算，可以得到$L_i$为2.9。

![screen shot 2018-01-22 at 15 34 13](https://user-images.githubusercontent.com/35488201/35309950-f489c998-00e9-11e8-9b56-645f6b2af549.png)

- 上图是我们对汽车和青蛙分别计算之后的结果带进公式中，我们得到最终的损失值为5.27.

#### Questions

1. 如果我们稍微改变汽车的分数，损失函数会发生什么变化么？
   - 如果汽车图像的分数稍微发生变化，损失函数将不会发生变化。SVM损失函数只关注于正确的分数比正确的分数大于1，但是这种情况下，汽车的分数比其他的分数都要大，所以如果汽车的分数改变了一丢丢，那么1的界限依旧有效，损失函数并不会改变。
2. 损失函数的最大最小只可能是多少？
   - 最小值为0.最大值为无穷大。
3. 当你初始化这些参数并且从头开始训练时，通常先使用一些很小的随机值来初始化W，你的分数的结果在训练的初期倾向于呈现较小的均匀分布的值。并且问题在于如果你所有的s，也就是你所有的分数都近乎于0，并且差不多相等，那么当你使用多分类SVM时，损失函数的预计会是多少？
   - 分类的数量减去1。因为如果我们对所有不正确的类别遍历上一遍，那么我们实际上遍历了C-1个类别，在这些类别中的每一个的分数差不多相等，所以我们就会得到一个值为1的损失值，所以我们的最终结果为C-1。这种方法十分有用，因为在刚开始调参的时候，你要知道你的损失函数是多少，如果在刚开始调参的时候，你的第一次迭代的损失函数并不等于C-1，这说明可能存在一个bug。
4. 如果我们使用求平均值而不是求和呢？
   - 不会改变。分类的数量需要提前确定，因为这只是将损失函数缩小了一个倍数，所以并没有什影响。
5. 如果我们在max里面加上平方，会怎么样？
   - 这会使损失函数变成另外一种函数。这里就说明我们用非线性改变了好坏之间的平衡，所以实际上我们使用了另外一种损失函数。

#### Example code

```python
def L_i_vectorized(x, y, W):
    scores = W.dot(x)
    margins = np.maximum(0, scores - scores[y] + 1)
    margins[y] = 0
    loss_i = np.sum(margins)
    return loss_i
```

#### 使损失函数为零的W是否唯一？

![screen shot 2018-01-22 at 18 42 09](https://user-images.githubusercontent.com/35488201/35309955-ff8af966-00e9-11e8-953d-701b2e73fefa.png)

- 加入我们找到了一个W，使得损失函数为零，那这个W是不是唯一的？

  - 不是唯一的，当我们把W乘以2时，损失函数也是零。

![screen shot 2018-01-22 at 18 46 22](https://user-images.githubusercontent.com/35488201/35309957-0530c2d8-00ea-11e8-9a7b-10b89ae459e8.png)

#### 损失器到底在这些不同的W之间是怎样做的？

- 我们训练出来分类起，更希望分类起在测试集上有更好的表现，如果让它仅仅去拟合训练集，这并不是我们想看到的。

![screen shot 2018-01-22 at 18 51 41](https://user-images.githubusercontent.com/35488201/35309960-0b64b1dc-00ea-11e8-890e-84de96edf900.png)

- 这并不是一个线性分类，而是一个更广义的机器学习内容。

- 图中的蓝点为训练集，并且我们训练出来了一条蓝色的曲线去拟合它。然后我们引入绿色的数据，这样，蓝色的曲线将不再是我们希望的曲线，我们更希望绿色的曲线是预测出来的曲线。为了解决这个问题，我们引入了正则化。

![screen shot 2018-01-22 at 18 56 17](https://user-images.githubusercontent.com/35488201/35309969-1125782c-00ea-11e8-9fe9-fe157cb00fbe.png)

- 我们在损失函数里加上一个附加项，我们称它为正则项，用于鼓励模型以某种方式选择更简单的W。这样标准的损失函数就拥有两项，数据损失项和正则损失项。而𝜆就用来平衡着两个项。（使模型简单化两种方法：1.对模型加以限制，防止模型过于复杂。2.加入惩罚项。）

### 正则化

- 我们希望能向某些特定的权重**W**添加一些偏好，对其他权重则不添加，以此来消除模糊性,这一点是能够实现的，方法是向损失函数增加一个**正则化惩罚（regularization penalty）**$R(W)$部分。正则化函数不是数据的函数，而是仅仅针对权重。

#### 常见的正则化函数

![screen shot 2018-01-22 at 19 10 35](https://user-images.githubusercontent.com/35488201/35309973-16adcfe2-00ea-11e8-8f42-5ad4260cb4ad.png)

1. L2正则化（欧式范数的权重W）：L2正则化是对这个权重响亮的欧式范数进行惩罚,它通过对所有参数进行逐元素的平方惩罚来抑制大数值的权重。
2. L1正则化
3. 弹性网络正则化
4. 最大规范正则化

- 正则化的目的是减轻模型的复杂度，而不是去试图拟合数据。

- 这里引入正则化最好的性质就是对大数值权重进行惩罚，可以提高其泛化能力，因为这就意味着没有哪个维度能够独自对于整体分值有过大的影响

![screen shot 2018-01-22 at 19 51 35](https://user-images.githubusercontent.com/35488201/35309979-1f28a19c-00ea-11e8-80d9-7a2cde71c2b2.png)

  - 举个例子，假设输入向量$x = [1,1,1,1]$，两个权重向量$w_1 = [1,0,0,0]$,$w_2 = [0.25,0.25,0.25,0.25]$。那么$w^T_1 x= w^T_2x = 1$，两个权重向量都得到同样的内积，但是$w_1$的L2惩罚是1.0，而$w_2$的L2惩罚是0.25。因此，根据L2惩罚来看，$w_2$更好，因为它的正则化损失更小。从直观上来看，这是因为的$w_2$权重值更小且更分散。既然L2惩罚倾向于更小更分散的权重向量，这就会鼓励分类器最终将所有维度上的特征都用起来，而不是强烈依赖其中少数几个维度。在后面的课程中可以看到，这一效果将会提升分类器的泛化能力，并避免*过拟合*。
  - 超参数∆和𝜆看起来是两个不同的超参数，但实际上他们一起控制同一个权衡：即损失函数中的数据损失和正则化损失之间的权衡

### Softmax Classifier（多项逻辑斯蒂回归）

- 在SVM分类中，我们并没有解释每一项预测的得分的真正含义，我们只是给出了这一项更可能是这一种分类，但对于softmax损失函数，我们将赋予这些得分一些额外的含义。我们将用这些得分，针对我们的类别，去计算概率分布。当谈论分数时，我们将softmax损失函数指数化以便结果都是正数，然后利用这些指数的和来归一化。当分数经过softmax函数处理之后，就得到了相应的概率，每个概率都分布在0和1之间，所有类别的概率和等于1。而我们希望对应相应的类别的概率更接近1。![Screen Shot 2018-01-22 at 20.06.47](/Users/zhouzhou/Desktop/Screen Shot 2018-01-22 at 20.06.47.png)
- 只要使概率最大，就能找到log的最大值，同时损失函数是用来度量坏的程度，所以我们加了一个负号。所以最后损失函数就变成了$- log P$.
- 在softmax分类器中，函数映射$f(x_i;W) = Wx_i$ 保持不变，但将这些评分值视为每个分类的未归一化的对数概率并且将*折叶损失（hinge loss）*替换为**交叉熵损失**（**cross-entropy loss）**。
- 其中函数$f_j(z) = {e^{z_j}}/\sum_k{e^{f_k}}$被称为**siftmax函数**，其输入值是一个向量，向量中元素为任意实数的评分值（z中的），函数对其进行压缩，输出一个向量，其中每个元素值在0到1之间，且所有元素之和为1。

<img width="900" alt="screen shot 2018-01-22 at 21 20 43" src="https://user-images.githubusercontent.com/35488201/35309986-28fac880-00ea-11e8-80a9-94bfd3ed4f05.png">

#### Example

![screen shot 2018-01-22 at 21 27 41](https://user-images.githubusercontent.com/35488201/35309989-2fbc32a8-00ea-11e8-8f6e-aa45e49181f6.png)

#### Questions

1. Sofatmax损失函数的最大值和最小值是多少？
   - 最小值是0，最大值是无穷大。对于分类问题，我们希望正确的概率是1，不正确的概率是0。这样的话，概率为1的$L_i$的取值就是0，而概率为0的取值就是无穷大。
2. 如果所有的s都接近于0，损失值是多少？
   - logC

### SVM线性分类和softnmax对比

![screen shot 2018-01-22 at 21 58 48](https://user-images.githubusercontent.com/35488201/35309994-3528bdd8-00ea-11e8-9900-6c1bf3572100.png)

- 两个损失函数需要面对的问题是如何解释这些分值，今儿量化度量到底有多坏。

![screen shot 2018-01-22 at 22 02 20](https://user-images.githubusercontent.com/35488201/35309996-39bacd64-00ea-11e8-9ac6-03d11bf7f98e.png)

  - 这里SVM分类并不会产生变化，前面已经讲过。
  - 但是对于softmax就会不同。softmax的目的是让概率质量函数（离散分布值）等于1，所以你给正确分类再高的分值，同时给不正确的分类再低的分值，softmax依旧会在正确的分类上积累更多的概率质量，正确的阶数向无穷大迈进。并持续降这一分值推向无穷大，并将不正确分类的分支推向负无穷大。
  - SVM对于超过某一阈值的数据点就不再处理，而softmax将会使每一个数据点越来越好。

### Recap

![screen shot 2018-01-22 at 22 12 44](https://user-images.githubusercontent.com/35488201/35310000-3f37ff32-00ea-11e8-9a39-485e43fa642c.png)

**Comment by wangchau** [comment]
<img width="969" alt="screen shot 2018-01-28 at 21 55 01" src="https://user-images.githubusercontent.com/35488201/35482999-ab758cd8-0477-11e8-8682-37f0c2d96631.png">

文中softmax的一处图片

**Comment by wangchau** [comment]
![screen shot 2018-01-22 at 20 06 47](https://user-images.githubusercontent.com/35488201/35483017-d2c523c0-0477-11e8-9cc2-91c8f9bf6c13.png)

文中softmax的一张图片
